### PR TITLE
Return the byte length of data by reference from the socket_recv* functions

### PR DESCRIPTION
**Pull request by mrrsm**
Takes into consideration the discussion that happened in #21

### CODE REVIEW

File: src/Socket.php [hunk 1]
```diff
@@ -255,13 +255,14 @@ public function read($length, $type = PHP_BINARY_READ)
...
      * @param int $flags
+     * @param int $ref reference will be filled with the length of data read from the socket
      * @return string
      * @throws Exception on error
      * @see self::read() if you do not need to pass $flags
      * @see self::recvFrom() if your socket is not connect()ed
      * @uses socket_recv()
      */
-    public function recv($length, $flags)
+    public function recv($length, $flags, &$ret)
```
Comment (mrrsm) [inline src/Socket.php]: I am not sure if you want me to set this to null by default (which would make it not break current implementations of it).

Comment (clue) [inline src/Socket.php]: :+1: for not breaking BC

**Comment by clue** [comment]
I very much appreciate the work you're putting into this, keep it up :+1: 

Admittedly, I'm still not sold on how useful this addition is.

However, given how little it affects the current API, I'm inclined this say this one is way *more likely* to get in…

Can you add some documentation, examples or test cases to highlight how using this additional parameter is beneficial?

Also, what's your take on the inconsistency between `read()`, `recv()` and `recvFrom()`?

**Comment by mrrsm** [comment]
>Also, what's your take on the inconsistency between read(), recv() and recvFrom()?

I prefer the consistent return of the data by all that this wrapper does.  It makes more sense to me. I feel like that should have been how they behaved in php to start with.  Instead of buffer being returned by reference it should be the length read thus making it optional and not using extra memory for it for those that don't need it (which is probably most of the people who are using it).

Changes from above incoming shortly.

**Comment by mrrsm** [comment]
>Can you add some documentation, examples or test cases to highlight how using this additional parameter is beneficial?

I am not sure about writing a test that would be useful based on what is already there without setting up a server to use to get the data from.

Here is an example, which I am willing to put into the README if you wish but I don't know if it is really useful for most people.

You have a message being received over the socket.  Its format is 4 byte unsigned int length followed by the data.

You get told to read 2147483648 bytes of data.

You could then use
```php
$read = 2147483648;

do {
    // Even with MSG_WAITALL the socket can return less then $read length of data
    $data .= $socket->recv($read, MSG_WAITALL, $length);
    $read -= $length;
} while($read > 0);
```

**Comment by clue** [comment]
> > Also, what's your take on the inconsistency between read(), recv() and recvFrom()?

> I prefer the consistent return of the data by all that this wrapper does.

Both `recv()` and `recvFrom()` now have an additional `&$ret` parameter, while `read()` does not.

I like the use case in your example. I suppose this could be something more people are interested in. However, I think this still looks quite cumbersome (perhaps even confusing), even with your additional parameters. FWIW this also doesn't take a closed connection into account (`$length === 0`).

Perhaps we could (instead?) offer this as an alternative method?

**Comment by mrrsm** [comment]
>Both recv() and recvFrom() now have an additional &$ret parameter, while read() does not.

That is also true for the php socket_[read,recv,recvfrom] functions as well though.  I feel like if you just needed the data read would be sufficient for most people as it just gets the data.

>FWIW this also doesn't take a closed connection into account ($length === 0).

You are correct, I simplified the full code to be more easily digestible to get my point across.  I also should be wrapping all of that in a try catch.

>I think this still looks quite cumbersome (perhaps even confusing)

Agreed, we have an odd use case compared to most people using sockets in php and this change is different from what is currently presented in the api.

>Perhaps we could (instead?) offer this as an alternative method?

Do you have any ideas on naming of the two functions?
If we do new functions for these I would prefer to keep it returning the socket read length and pass buffer by reference.  That way it is a direct wrapper for the socket functions and you have the current implementation for getting it the other way.

**Comment by clue** [comment]
I appreciate the work you've put into this and would like to thank you for your PR :+1: 

However, after reconsidering this change further, I'm having to reject this PR:

* This lib aims to be lightweight, simple wrapper around `ext-sockets` and aims to hide some of its inconsistencies / quirks (from a consumer's perspective)
* The `$ret` parameter is non-obvious and of little use for most use cases. I would argue it's hard to understand and perhaps even misleading
* `strlen()` is the accepted and well-understood alternative
* There appears to be little interest in this

>> Perhaps we could (instead?) offer this as an alternative method?

> Do you have any ideas on naming of the two functions?
If we do new functions for these I would prefer to keep it returning the socket read length and pass buffer by reference. That way it is a direct wrapper for the socket functions and you have the current implementation for getting it the other way.

I'm interested in seeing how well this fits the current design :+1: Let's just start with a `readAll()` (or similar) and discuss this in a separate PR?

Again, I would like to thank you for your contribution and welcome and encourage further contributions, keep it up :+1:

**Pull request closed without merging**